subworkflows/local/eggnog.nf

```diff
@@ -0,0 +1,40 @@
+//
+// Run eggnog-mapper on called ORFs, first optionally downloading the required databases
+//
+
+include { EGGNOG_DOWNLOAD } from '../../modules/local/eggnog/download'
+include { EGGNOG_MAPPER   } from '../../modules/local/eggnog/mapper'
+include { EGGNOG_SUM      } from '../../modules/local/eggnog/sum'
+
+workflow EGGNOG {
+    take:
+    eggnog_dbpath
+    faa
+    collect_fcs
+
+    main:
+    ch_versions = Channel.empty()
+    String eggnogDB      = eggnog_dbpath + "/eggnog.db"
+    File eggnogfile      = new File(eggnogDB)
+
+    if ( ! eggnogfile.exists() ) {
+        EGGNOG_DOWNLOAD()
+        ch_dbpath = EGGNOG_DOWNLOAD.out.db
+        ch_versions = ch_versions.mix ( EGGNOG_DOWNLOAD.out.versions )
+    } else {
+        ch_dbpath = Channel.fromPath(eggnog_dbpath, checkIfExists: true)
+    }
```

Comment:
I get error when I use `ch_dbpath = EGGNOG_DOWNLOAD.out.db`
mainly because `eggnog` checks whether the files are already stored and return an error if they are present. 
So I don't know how to avoid that!

Reply:
So in a scenario where I need to download the database it works smoothly but when I give as input my own folder, I get the error.

Reply:
> I get error when I use `ch_dbpath = EGGNOG_DOWNLOAD.out.db` mainly because `eggnog` checks whether the files are already stored and return an error if they are present. So I don't know how to avoid that!

Could you clarify this part please? What error is returned? What are your parameters?

Reply:
```
     input:
     path "eggnog"
 
     output:
     path "$eggnog"          , emit: dirdb
     path("$eggnog/*db") , emit: db
     path "versions.yml"    , emit: versions
 
     script:
     def args = task.ext.args ?: ''
     """
 
     download_eggnog_data.py \\
         $args \\
         -y \\
         --data_dir $eggnog
```

this is how the eggnog module looks like

Reply:
What is the need for both `dirdb` and `db`? I'm definitely confused by the usage of EGGNOG here. 

The implementation pattern should look like:
```nextflow
params.user_supplied_db = null //
params.database_name = 'default'

workflow {
    if( !params.user_supplied_db ){
        DOWNLOAD_DB( params.database_name )
    }
    TOOL(
         ch_input,     // [ meta, input ]
         params.user_supplied_db ? file(params.user_supplied_db, checkIfExists:true): DOWNLOAD_DB.out.db // db
    )
}
```

So are you passing the user supplied database to the download process?

Reply:
Yes I do. EGGNOG needs to download into a directory. But if the files are already in the directory, you'll get an error.

Reply:
Why not avoid passing an input and just do `mkdir eggnog` in the script? That way it will always be empty.
Using the pattern above, the download process doesn't run if the user provides an existing database path, which seems to be the behavior you want.

Reply:
isn't it a similar workaround that we have now? then it will be with directory name instead of filename. the eggnog_download´module had ´mkdir eggnog´indeed.

Reply:
The intention of the original comment was to remove the need of
```
    String eggnogDB      = eggnog_dbpath + "/eggnog.db"
    File eggnogfile      = new File(eggnogDB)

    if ( ! eggnogfile.exists() ) {
```
part specifically.
So yes the strategies are similar, but this can be handled without creating a new File object. 
This method doesn't notify the user if that the expected eggnog.db database is not in the place they expect. It just runs the download process instead.